aws_lambda_powertools/utilities/streaming/s3_object.py

```diff
@@ -0,0 +1,235 @@
+)
+from aws_lambda_powertools.utilities.streaming.transformations.base import (
+    BaseTransform,
+    T,
+)
+
+if TYPE_CHECKING:
+    from mypy_boto3_s3 import S3ServiceResource
+
+
+class S3Object(IO[bytes]):
+    """
+    Seekable streamable S3 Object reader.
+
+    S3Object implements the IO[bytes], backed by a seekable s3 streaming.
+
+    Parameters
+    ----------
+    bucket: str
+        The S3 bucket
+    key: str
+        The S3 key
+    version_id: str, optional
+        A version ID of the object, when the S3 bucket is versioned
+    boto3_s3_resource: S3ServiceResource, optional
+        An optional boto3 S3 resource. If missing, a new one will be created.
+    gunzip: bool, optional
+        Enables the Gunzip data transformation
+    csv: bool, optional
+        Enables the CSV data transformation
+
+    Example
+    -------
+
+    ** Reads a line from an S3, loading as little data as necessary
+
+        >>> from aws_lambda_powertools.utilities.streaming import S3Object
+        >>>
+        >>> line: bytes = S3Object(bucket="bucket", key="key").readline()
+        >>>
+        >>> print(line)
+
+    """
+
+    def __init__(
+        self,
+        bucket: str,
+        key: str,
+        version_id: Optional[str] = None,
+        boto3_s3_resource: Optional["S3ServiceResource"] = None,
+        gunzip: Optional[bool] = False,
+        csv: Optional[bool] = False,
+    ):
+        self.bucket = bucket
+        self.key = key
+        self.version_id = version_id
+
+        # The underlying seekable IO, where all the magic happens
+        self.raw_stream = _S3SeekableIO(
+            bucket=bucket, key=key, version_id=version_id, boto3_s3_resource=boto3_s3_resource
+        )
+
+        # Stores the list of data transformations
+        self._data_transformations: List[BaseTransform] = []
+        if gunzip:
+            self._data_transformations.append(GzipTransform())
+        if csv:
+            self._data_transformations.append(CsvTransform())
+
+        # Stores the cached transformed stream
+        self._transformed_stream: Optional[IO[bytes]] = None
+
+    @property
+    def size(self) -> int:
+        """
+        Retrieves the size of the underlying S3 object
+        """
+        return self.raw_stream.size
+
+    @property
+    def transformed_stream(self) -> IO[bytes]:
+        """
+        Returns a IO[bytes] stream with all the data transformations applied in order
+        """
+        if self._transformed_stream is None:
+            # Apply all the transformations
+            transformed_stream = self.raw_stream
+            for transformation in self._data_transformations:
+                transformed_stream = transformation.transform(transformed_stream)
+
+            self._transformed_stream = transformed_stream
+
+        return self._transformed_stream
+
+    @overload
+    def transform(
+        self, transformations: Union[BaseTransform[T], Sequence[BaseTransform[T]]], in_place: Literal[True]
```

Comment:
In 3.7, we can now use `|` instead of Union if you prefer.

```
transformations: `BaseTransform[T] | Sequence[BaseTransform[T]`
```

you need an import per module you want to do that:

```python
from __future__ import annotations
```